File: Arrays and String/54. Spiral Matrix/spiral.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        spiral = []

        M = len(matrix)
        N = len(matrix[0])

        left_col = 0
        right_col = N - 1
        top_row = 0
        bot_row = M - 1

        while top_row <= bot_row and left_col <= right_col:
            for j in range(left_col, right_col + 1):
                spiral.append(matrix[top_row][j])
            top_row += 1

            if top_row > bot_row or left_col > right_col:
                break

            for i in range(top_row, bot_row + 1):
                spiral.append(matrix[i][right_col])
            right_col -= 1

            if top_row > bot_row or left_col > right_col:
                break

            for j in range(right_col, left_col - 1, -1):
                spiral.append(matrix[bot_row][j])
            bot_row -= 1

            if top_row > bot_row or left_col > right_col:
                break

            for i in range(bot_row, top_row - 1, -1):
                spiral.append(matrix[i][left_col])
            left_col += 1

        return spiral

File: 3. Arrays and String/54. Spiral Matrix/spiral.cpp
class Solution
{
public:
    vector<int> spiralOrder(vector<vector<int>> &matrix)
    {
        vector<int> spiral;

        int M = matrix.size();
        int N = matrix[0].size();

        int left_col = 0;
        int right_col = N - 1;
        int top_row = 0;
        int bot_row = M - 1;

        while (top_row <= bot_row && left_col <= right_col)
        {
            for (int j = left_col; j <= right_col; j++)
            {
                spiral.push_back(matrix[top_row][j]);
            }
            top_row++;

            if (top_row > bot_row || left_col > right_col)
            {
                break;
            }

            for (int i = top_row; i <= bot_row; i++)
            {
                spiral.push_back(matrix[i][right_col]);
            }
            right_col--;

            if (top_row > bot_row || left_col > right_col)
            {
                break;
            }

            for (int j = right_col; j >= left_col; j--)
            {
                spiral.push_back(matrix[bot_row][j]);
            }
            bot_row--;

            if (top_row > bot_row || left_col > right_col)
            {
                break;
            }

            for (int i = bot_row; i >= top_row; i--)
            {
                spiral.push_back(matrix[i][left_col]);
            }
            left_col++;
        }
        return spiral;
    }
};